packages/components/tabs/src/useResizeObserver.ts

```diff
@@ -0,0 +1,42 @@
+import { type RefObject, useEffect, useRef, useState } from 'react'
+
+interface Size {
+  width?: number
+  height?: number
+}
+
+type ResizeCallback = (entry?: ResizeObserverEntry) => void
+
+export const useResizeObserver = <T extends HTMLElement>(
+  target: RefObject<T> | T | null,
+  onResize?: ResizeCallback
+): Size => {
+  const [size, setSize] = useState<Size>({ width: undefined, height: undefined })
+  const resizeObserverRef = useRef<ResizeObserver>()
+  const resizeCallbackRef = useRef<ResizeCallback | undefined>(onResize)
+
+  useEffect(() => {
+    const targetElm = target && 'current' in target ? target.current : target
+    if (!targetElm) {
+      return
+    }
+
+    if (!resizeObserverRef.current) {
```

Comment:
to avoid a nesting block, maybe switch the condition:
`if (resizeObserverRef.current) return`

Reply:
My point was to switch the predicate, not so much the syntax ;)
```tsx
if (resizeObserverRef.current) {
  return
}
```

Reply:
This is also not allowed... And I'd have to `return undefined`, which seems uglier than current option, don't you think?

Reply:
Hmmm... My bad, it seems I was mistaking here :thinking:

Reply:
Hmmm sorry I was in a mistake here with the `undefined`. Anyway, the thing is here, I dont' want to exit from function if `resizeObserverRef.current` is defined... I just want here to create a new `ResizeObserver` in case there isn't already one.

Reply:
> I just want here to create a new ResizeObserver in case there isn't already one.

Yes, so you should exit the function if there is already a set resizeObserver to prevent creating a new one, right?

Reply:
If there is already a `resizeObserverRef`, I will use it, not exit the function: https://github.com/adevinta/spark/blob/544-components-tabs/packages/components/tabs/src/useResizeObserver.ts#L33

Reply:
```tsx
useEffect(() => {
    if (resizeObserverRef.current) return

    const targetElm = target && 'current' in target ? target.current : target
    if (!targetElm) {
      return
    }

    resizeObserverRef.current = new ResizeObserver(([entry]) => {
      const { inlineSize: width, blockSize: height } = entry?.borderBoxSize[0] ?? {}
      resizeCallbackRef.current?.(entry)

      setSize({ width, height })
    })

    resizeObserverRef.current.observe(targetElm as unknown as HTMLElement)

    return ...
  }, [target, resizeCallbackRef])
```

The code above checks if a resizeObserver is already set, and if so, returns. If not, a new ResizeObserver is created, and its observe method is called on the target element.

vs
```tsx
useEffect(() => {
    const targetElm = target && 'current' in target ? target.current : target
    if (!targetElm) {
      return
    }

    if (!resizeObserverRef.current) {
      resizeObserverRef.current = new ResizeObserver(([entry]) => {
        const { inlineSize: width, blockSize: height } = entry?.borderBoxSize[0] ?? {}
        resizeCallbackRef.current?.(entry)

        setSize({ width, height })
      })
    }

    resizeObserverRef.current.observe(targetElm as unknown as HTMLElement)

    return () => {
      resizeObserverRef.current &&
        resizeObserverRef.current.unobserve(targetElm as unknown as HTMLElement)
    }
  }, [target, resizeCallbackRef])
```

The code above checks if there is no ResizeObserver already set, then create a new one. However, it will always call the `observe` method on it, even if it's not needed.

Reply:
Aaaaaaaaaaaaah yes I get it, sorry :sweat_smile: